tests/optimize/test_qnspsa.py

```diff
@@ -0,0 +1,229 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for the ``QNSPSAOptimizer``
+"""
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+def get_single_input_qnode():
+    dev = qml.device("default.qubit", wires=2)
+    # the analytical expression of the qnode goes as:
+    # np.cos(params[0][0] / 2) ** 2 - np.sin(params[0][0] / 2) ** 2 * np.cos(params[0][1])
+    @qml.qnode(dev)
+    def loss_fn(params):
+        qml.RY(params[0][0], wires=0)
+        qml.CRX(params[0][1], wires=[0, 1])
+        return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+    return loss_fn, (2, 2)  # returns the qnode and the input param shape
+
+
+def get_multi_input_qnode():
+    dev = qml.device("default.qubit", wires=2)
+    # the analytical expression of the qnode goes as:
+    # np.cos(x1 / 2) ** 2 - np.sin(x1 / 2) ** 2 * np.cos(x2)
+    @qml.qnode(dev)
+    def loss_fn(x1, x2):
+        qml.RY(x1, wires=0)
+        qml.CRX(x2, wires=[0, 1])
+        return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+    return loss_fn
+
+
+class TestSPSAOptimizer:
```

Comment:
The current tests are looking good now 💯 .

Could you also add integration tests that use the optimizer `step` and `step_and_cost` methods directly? Should be relatively straightforward from the ones you currently have.